src/libraries/ButtonHandler/ButtonHandler.cpp

```diff
@@ -0,0 +1,144 @@
+#include <Arduino.h>
+#include <ButtonHandler.h>
+
+ButtonHandler * _buttonHandler;
+
+// Empty constructor
+ButtonHandler::ButtonHandler(ButtonConfig config) : _config(config)
+{
+    // Global reference needed for global ISR pointer
+    _buttonHandler = this;
+
+    for (int i = 0; i < ButtonConfig::MAX_BUTTONS; i++) 
+    {   
+        if (_config.buttonModes[i] != DISCONNECTED) {
+            pinMode(_config.buttonPins[i], INPUT);
+            attachInterrupt(_config.buttonPins[i], getISR(i), RISING);
+
+            // Only attach falling edge interrupts for those keys we care about holding down, i.e modifiers
+            if (config.buttonModes[i] == MODIFIER) {
+                attachInterrupt(_config.buttonPins[i], getISR(i), FALLING);
+            }
+        }
+    }
+}
+
+// Setting the callback function used to emit events to
+void ButtonHandler::setListener(EventListener listener) {
+    _listener = listener;
+}
+
+// Emits the next queued event if it exists
+void ButtonHandler::eventComplete() 
+{
+    
+    if (!isEventsEmpty()) {
+        _listener(_events[0]);
+        resizeEvents();
+    } else {
+        _processingComplete = true;
+    }
+}
+
+// Returns whether or not the event queue is empty
+boolean ButtonHandler::isEventsEmpty() 
+{
+    return _lastIndex == 0;
+}
+
+// Shifts event array down an indices such that index 1 -> index 0 etc.
+void ButtonHandler::resizeEvents() 
+{
+    for (int i = 0; i < _lastIndex; i++) {
+        _events[i] = _events[i+1];
+    }
+
+    _lastIndex -= 1;
```

Comment:
I think from C paradigms, `--` is apparently not obvious? I remember reading that somewhere. Anyway, I don't think we're gunna ever need to increment a variable by more than 1 at a time, so will change to `++` and `--`